woocommerce/woocommerce
# Deprecate the $check_key_exists parameter from AssetDataRegistry and disallow duplicate data for all cases

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Currently, [AssetDataRegistry::add_data](https://github.com/woocommerce/woocommerce/blob/25d42b0c053c0b12db7401b5ec33b7c760a81fda/plugins/woocommerce/src/Blocks/Assets/AssetDataRegistry.php#L429) doesn't allow overriding existing data for a given key. 

Still, the only method that calls it, [AssetDataRegistry::add](https://github.com/woocommerce/woocommerce/blob/25d42b0c053c0b12db7401b5ec33b7c760a81fda/plugins/woocommerce/src/Blocks/Assets/AssetDataRegistry.php#L305), has an optional param `$check_key_exists` with the default value set to false, which in practice doesn't have any impact in the end results, as whenever set to false, no data is overridden: in this case we have an early return combined with a fatal error triggered exclusively whenever `WP_DEBUG` is set to TRUE.

This logic has caused us a few problems, and after internal discussion, which happened on pdnLyh-5U0-p2, **I'm proposing for us to deprecate the `$check_key_exists` parameter**, as if we allow data override, quoting @senadir : 

> The structure of that data is considered public API, allowing it to be overridden opens the risk to having mismatched expectations between frontend code that’s consuming it and what was modified with. This is magnified by the fact that we evolved from passing simple primitives to objects in wcSettings.

Porting over @samueljseay 's comment here as well:

> I also had a recent discussion in slack p1708450555264569/1708435510.631329-slack-C03CPM3UXDJ about this. In almost all cases we pass true to check key exists or we guard by checking if it exists before adding it. So I think we should as you say remove the param and always check or set the default to true. We could warn if the key exists, because it might be an indication your code is not working as expected?

Following Sam's suggestion, I've also updated the pre-existing logic that would throw a fatal exclusively whenever DEBUG is set to true to throw a PHP warning instead.

This is a change that should not have any impact on functionality, still, I'm tagging folks from the various teams to review this PR and test so we can confirm things are working as expected.





Closes #43404



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure your WordPress install has `WP_DEBUG` [set to true](https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/).
2. Access WooCommerce > Home > Customize Your Store > Start designing ( `wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store%2Fassembler-hub` )
3. Make sure you can access the pattern assembler without any problems: no fatal errors should be triggered.
4. Do a general smoke test including the Templates and the Cart, Mini-Cart, Checkout, Product Image, Product Collection, Reviews, and Filter blocks, and confirm that everything is working as expected: you should not see any warnings in your PHP error log.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Deprecate the $check_key_exists parameter from AssetDataRegistry and disallow duplicate data for all cases.

#### Comment 

</details>
